src/helpers/defineMessagesAnalyzator.ts

```diff
@@ -0,0 +1,114 @@
+import { Rule } from "eslint";
+import * as ESTree from 'estree';
+import { Dictionary, CallExpressionNode } from "./types";
+
+type TrackedMessage = {
+  isReported: boolean,
+  node: ESTree.Property,
+  context: Rule.RuleContext,
+}
+
+type MessageId = string | number;
+
+export default class DefineMessagesDuplicationAnalyzator {
+  private idTracker: Dictionary<MessageId, TrackedMessage> = {};
+  private context?: Rule.RuleContext;
+
+  public setContext = (context: Rule.RuleContext): void => {
+    this.context = context;
+  }
+
+  private reportDuplication = (
+    messageId: MessageId, context: Rule.RuleContext, messageNode: ESTree.Property
+  ): void  => {
+    context.report({
+      message: `message with id '${messageId}' is duplicated`,
+      node: messageNode,
+    });
+  }
+
+  private getMessageId = (messageNode: ESTree.Property): number | string | null => {
+    const messageNodeProperies: ESTree.Property[] = messageNode.value.type === 'ObjectExpression' ?
+      this.removeSpreadElements(messageNode.value.properties) : [];
+
+    const messageIdNode = messageNodeProperies.find((messagesProperty: ESTree.Property) => messagesProperty.key.type === 'Identifier' && messagesProperty.key.name === 'id');
+
+    if ((messageIdNode == null) || messageIdNode.value.type !== 'Literal') {
+      return null;
+    }
+
+    const messageId = messageIdNode.value.value;
+    if (typeof messageId === 'number' || typeof messageId === 'string') {
+      return messageId;
+    }
+
+    return null;
+  }
+
+  private checkMessageDuplication = (messageNode: ESTree.Property): void => {
+    const messageId: string | number | null = this.getMessageId(messageNode);
+    if (!messageId || (this.context == null)) {
+      return;
+    }
+
+    const isDuplication = !(this.idTracker[messageId] == null);
+    if (!isDuplication) {
+      this.idTracker[messageId] = {
+        node: messageNode,
+        context: this.context,
+        isReported: false,
+      };
+
+      return;
+    }
+
+    this.reportDuplication(messageId, this.context, messageNode);
+
+    const firstTrakedMessage = this.idTracker[messageId];
+    if ((firstTrakedMessage != null) && !firstTrakedMessage.isReported) {
```

Comment:
Then you should not need the `firstTrakedMessage != null` here

Reply:
If I understand correctly, the 'isReported' really matters if there is more the same id is in more than 2 places. Then, you should also add a test case with 3 messages that have the same id.

Reply:
Just moving wouldn't be enoght. As far I use `isDuplication` variable in condition check, the TS does not understand, that I alredy have checked `firstTrakedMessage`.

I will remove `isDuplication` variable. Hope it will not have big impackt on readability.

Reply:
I think it increased readability as we no longer have the double-negation.